Improvements to PyQGIS code testing guidelines

SrNetoChan opened the pull request:
### Description
<!---
Include a few sentences describing the overall goals for this Pull Request.
--->

Improvements to the code testing guidelines. It changes and separate some of the concepts, which, IMHO, makes the sequence more easy to follow.

Goal:

<!---
If your PR fixes a ticket, add `fix` in front of the ticket number. The ticket will be closed automatically.
Add as much as needed `fix #number` if the PR closes more than one ticket.
If your PR doesn't fix entirely the ticket, just add the ticket reference.
The complete list of the issues is at https://github.com/qgis/QGIS-Documentation/issues.
-->
Ticket: #

<!---
Indicate whether the fix should be backported to previous release.
Replace the space between square brackets by a `x` to make it checked.
-->
- [x] Backport to LTR documentation is required

### Minimal requirements for merging *(for maintainers)*
<!---
Reviewing is a process done by community members, mostly on a volunteer basis.
We try to keep the overhead as small as possible and appreciate if you help us.
In order for your PR to get merged it should satisfy some minimal requirements:
--->

- [x] The description of this PR is coherent with the manual and does not provide wrong information.
- [x] This PR passes the checks. <!---The results will be reported by travis-ci **after** opening the PR.-->

<!---
Please read carefully our writing guidelines at https://docs.qgis.org/testing/en/docs/documentation_guidelines/index.html
to help you fulfill these requirements. Feel free to ask in a comment if you have troubles with any of them.
--->

DelazJ commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -22,103 +24,127 @@ How to write testable code snippets
+    make -f docker.mk doctest
+
+  You have to install `Docker <https://www.docker.com/>`_ first because we will
+  use a docker image with QGIS in it.
```
Is it enough to install docker or should the user do more (eg, set a QGIS image in it or whatever the vocabulary/step is)?

SrNetoChan replied:
I haven't tried this, but for me it didn't make sense to have this sentence in the begining of the instructions when there are ways of runing the tests without docker. @ghtmtt probably have tested it. I will try to test it too.

ghtmtt replied:
you have to have docker installed and depending on your local qgis installation (compiled, package...) change the `user.mk` file in order to set up the correct environment

DelazJ commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -22,103 +24,127 @@ How to write testable code snippets
 Writing testable code snippets is not so different from the *old* method.
```
Because I hate section with a single subsection...
```suggestion

Use appropriate Sphinx directive
--------------------------------

Writing testable code snippets is not so different from the *old* method.
```

SrNetoChan replied:
I am ok having a subtitle. Maybe `Doctest directives` ?

SrNetoChan replied:
Created a subsection called "Doctest sphinx directives"

DelazJ commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -1,19 +1,21 @@
+:ref:`PyQGIS-Developer-Cookbook`, then you should follow some rules to enable
+automatic testing of the code snippets.
 
 Testing is really important because allows checking automatically the quality of
```
```suggestion
Testing is really important because it allows checking automatically the quality of
```

havatv commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -22,103 +24,127 @@ How to write testable code snippets
+  .. testcode::
 
-.. code-block:: python
+      print("QGIS CRS ID:", crs.srsid())
```
```suggestion
      print("QGIS CRS ID:", str(crs.srsid()))
```

SrNetoChan replied:
I just copied the old code, but, why use str() in here?

havatv replied:
srsid() returns a long integer.
```
python3
>>> print("CRS" + 5)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: must be str, not int
>>> print("CRS" + str(5))
CRS5
```

SrNetoChan replied:
I still don't see the need to had more code, the title to this is to show the attributes of CRS if the attribute is an integer, so be it, it's printed anyway. IMHO

havatv replied:
If it works, it works, but as you can see from the code that I included above, in my Python 3, not using str() makes it fail.

https://qgis.org/api/classQgsCoordinateReferenceSystem.html:

```
long srsid () const
     Returns the internal CRS ID, if available.
```

SrNetoChan replied:
Ok, I will test it in the console. thanks.

SrNetoChan replied:
I tested on QGIS console, which is using Python 3.6 and it printed fine. Not sure what is the catch here...

havatv replied:
Strange.  This is what I get in the Python console of QGIS 3.4.5 (Ubuntu 18.04):

```
print("CRS" + str(5))
CRS5
print("CRS" + 5)
Traceback (most recent call last):
  File "/usr/lib/python3.6/code.py", line 91, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
TypeError: must be str, not int
```

SrNetoChan replied:
I was not testing summing the "CRS" + 5

havatv replied:
Still in QGIS 3.4.5 (Ubuntu 18.04), Python console:
```
crs = QgsCoordinateReferenceSystem()
crs
<qgis._core.QgsCoordinateReferenceSystem object at 0x7fc0511f8798>
print("CRS" + crs.srsid())
Traceback (most recent call last):
  File "/usr/lib/python3.6/code.py", line 91, in runcode
    exec(code, self.locals)
  File "<input>", line 1, in <module>
TypeError: must be str, not int
print("CRS" + str(crs.srsid()))
CRS0
```

havatv commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -22,103 +24,127 @@ How to write testable code snippets
 
-.. code-block:: python
+      print("QGIS CRS ID:", crs.srsid())
+      print("PostGIS SRID:", crs.srid())
```
```suggestion
      print("PostGIS SRID:", str(crs.srid()))
```

SrNetoChan replied:
again

SrNetoChan replied:
this postgis example was wrong, I have updated it.

havatv commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -22,103 +24,127 @@ How to write testable code snippets
+      PostGIS SRID: 4326
 
-and then adding the expected results::
+Grouping code snippets tests
```
```suggestion
Grouping tests
```

havatv commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -22,103 +24,127 @@ How to write testable code snippets
-    .. testoutput:: mysnippet
+For each rst document, the code snippets are tested sequentially, which means
+you can use one ``.. testsetup::`` for all the following code snippets and that
+latter snippets will use variables declared in previous ones.
```
```suggestion
latter snippets can use variables declared in previous ones.
```

SrNetoChan replied:
The importance of `will` is that if you forget to clean up the variable it will be reused (happend to me in the CRS section). But maybe there is a better way to write that.

havatv replied:
It was implied by the suggested text.  I have updated with a new proposal.

SrNetoChan commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -22,103 +24,127 @@ How to write testable code snippets
+something like ``assert object.isValid()``) you can still compare the output
+with another directive using the ``print()`` method. If your code has printed
+outputs, you need to add the expected results within a ``.. testoutput::`` to
+compare the expected output::
```
They are, but they can be hidden if we want. You are right, it's a good idea to include that information.

SrNetoChan replied:
They are visible, but we can also make hide them if we want. I will try to had that

SrNetoChan replied:
Done

SrNetoChan commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -22,103 +24,127 @@ How to write testable code snippets
+   crs_fromwkt. In case of failures, this will help identifying where the failures
+   occur.
+
+If you don't declare any group, the code snippet will be added to the *default*
```
It really a group called "default"

SrNetoChan replied:
I have reworded this

SrNetoChan commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -22,103 +24,127 @@ How to write testable code snippets
-    .. testoutput:: mysnippet
+For each rst document, the code snippets are tested sequentially, which means
+you can use one ``.. testsetup::`` for all the following code snippets and that
+latter snippets will use variables declared in previous ones.
```
Both

SrNetoChan commented:
I think I have addressed all comments. I would like to hear @ghtmtt opinion before I merge it. After all he was the original author of the base document,

ghtmtt commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -1,19 +1,21 @@
-Testing is really important because allows checking automatically the quality of
-the code. During testing, if the code has errors, tests will fail, meaning that
-the code snippet has errors and needs to be fixed.
+Testing is really important because it allows automatic checking of
```
```suggestion
Testing is really important because allows automatic checking of
```

ghtmtt commented on source/docs/documentation_guidelines/cookbook_guidelines.rst:
```diff
@@ -1,19 +1,21 @@
-the code snippet has errors and needs to be fixed.
+Testing is really important because it allows automatic checking of
+the code. Code snippets with errors or code that uses outdated methods will fail and
+the notification will help you fix the problems.
```
```suggestion
the notification will help you to fix the problems.
```

ghtmtt replied:
not sure here

havatv replied:
Leave it as it is.
